nncf/quantization/algorithms/accuracy_control/ranker.py

```diff
@@ -0,0 +1,301 @@
+from abc import abstractmethod
+from typing import Callable, Any, List, Iterable, Optional, Iterable
+from dataclasses import dataclass
+
+from nncf.data.dataset import Dataset
+from nncf.common.factory import EngineFactory
+from nncf.common.graph import NNCFGraph
+from nncf.common.graph import NNCFNode
+from nncf.common.quantization.quantizer_removal import find_quantizer_nodes_to_cut
+from nncf.common.quantization.quantizer_removal import revert_operations_to_floating_point_precision
+from nncf.quantization.algorithms.accuracy_control.backend import AccuracyControlAlgoBackend
+
+
+def get_ranking_subset_indices(errors: List[float], ranking_subset_size: int) -> List[int]:
+    """
+    Returns `ranking_subset_size` indices of elements in the `errors` list
+    that have the biggest error value. Returned indices are sorted in
+    ascending order.
+
+    :param errors: A list of errors.
+    :param ranking_subset_size: A number of returned indices.
+    :return: Indices of elements in the `errors` list which have the biggest error value.
+    """
+    ordered_indices = [
+        idx for idx, _ in sorted(enumerate(errors), key=operator.itemgetter(1), reverse=True)
+    ]
+    end_index = min(ranking_subset_size, len(ordered_indices))
+    return sorted(ordered_indices[:end_index])
+
+
+@dataclass
+class GroupToRank:
+    """
+    Describes a group of quantizers to rank.
+
+    :param quantizers: List of quantizer nodes.
+    :param operations: List of nodes that will be reverted to original precision
+        if `GroupToRank.quantizers` are removed.
+    """
+
+    quantizers: List[NNCFNode]
+    operations: List[NNCFNode]
+
+    # TODO(andrey-churkin): Add hash
+
+
+class Ranker(ABC):
+    """
+    Encapsulates logic to rank groups of quantizers.
+    """
+
+    def __init__(self,
+                 ranking_subset_size: int,
+                 ranking_fn: Callable[[Any, Any], float],
+                 dataset: Dataset,
+                 algo_backend: AccuracyControlAlgoBackend):
+        """
+        :param ranking_subset_size: The number of data items that will be selected from
+            the dataset to rank groups of quantizers. The `len(dataset)` data items will
+            be selected if `ranking_subset_size` parameter is greater than the number of
+            elements in the dataset.
+        :param ranking_fn:
+        :param dataset: Dataset for the ranking process.
+        :param algo_backend: The `AccuracyControlAlgoBackend` algo backend.
+        """
+        self._ranking_subset_size = ranking_subset_size
+        self._ranking_fn = ranking_fn
+        self._dataset = dataset
+        self._algo_backend = algo_backend
+        # We don't need to re-calculate values for the initial model
+        # because they don't change. So use this attribute to store
+        # them to improve execution time.
+        self._ref_values = None
+
+    def find_groups_of_quantizers_to_rank(self, nncf_graph: NNCFGraph) -> List[GroupToRank]:
```

Comment:
Wrong method's parameters in the description.

Reply:
Done